### PR TITLE
DM ProcessState caching

### PR DESCRIPTION
**Pull request by arkadius**

### CODE REVIEW

File: docs/Changelog.md [hunk 1]
```diff
@@ -20,10 +20,13 @@
 * [#4101](https://github.com/TouK/nussknacker/pull/4101) Notifications fixes:
   * Scenario state wasn't refreshed after deploy/cancel action was successfully finished (was only after failure)
   * Notification "Deployment of ... failed ..." was presented even for cancel action 
-* [#4100](https://github.com/TouK/nussknacker/pull/4100) "In progress" state is now presented not only inside scenario editor but also on scenarios list
 * [#4102](https://github.com/TouK/nussknacker/pull/4102) Flink deploy now wait until job is started on TaskManagers before reporting that is finished - 
   thanks to that status and versions panel are refreshed with "DEPLOYED" state in the same time
 * [#3992](https://github.com/TouK/nussknacker/pull/3992) Fix for compiling scenarios containing filter node with only 'false' edge
+* [#4100](https://github.com/TouK/nussknacker/pull/4100)[#4104](https://github.com/TouK/nussknacker/pull/4104)
+  Before the change, in scenario list, were presented "local" scenario states - based only on Nussknacker's actions log.
```
Comment (jedrz) [inline docs/Changelog.md:27]: Maybe rephrase to?
> Before the change, the scenario list presented "local" states - based only on Nussknacker's actions log.

**Review by jedrz — APPROVED** [review]

File: docs/Changelog.md (hunk 1 above)
Comment (arkadius) [reply docs/Changelog.md]: It looks much simpler and better. Thanks!

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DeploymentManager.scala [hunk 2]
```diff
@@ -41,3 +42,13 @@ trait DeploymentManager extends AutoCloseable {
   def stop(name: ProcessName, savepointDir: Option[String], user: User): Future[SavepointResult]
 
 }
+
+trait AlwaysFreshProcessState { self: DeploymentManager =>
+
+  def getProcessState(name: ProcessName)
```
Comment (gadomsky) [inline designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DeploymentManager.scala:48]: can we make this method final?

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/cache/CachingProcessStateDeploymentManager.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import scala.concurrent.ExecutionContext.Implicits._
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+class CachingProcessStateDeploymentManager(delegate: DeploymentManager,
+                                           cacheTTL: FiniteDuration) extends DeploymentManager {
+
+  private val cache: AsyncCache[ProcessName, Option[ProcessState]] = Caffeine.newBuilder()
+    .expireAfterWrite(java.time.Duration.ofMillis(cacheTTL.toMillis))
+    .buildAsync[ProcessName, Option[ProcessState]]
```
Comment (gadomsky) [inline designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/cache/CachingProcessStateDeploymentManager.scala:24]: do we want to also cofigure max size of cache?

**Review by gadomsky — APPROVED** [review]
LGTM

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala [hunk 4]
```diff
@@ -272,11 +266,13 @@ class DeploymentServiceImpl(dispatcher: DeploymentManagerDispatcher,
   }
 
   private def getStateFromEngine(deploymentManager: DeploymentManager, processIdWithName: ProcessIdWithName)
-                                (implicit ec: ExecutionContext): Future[Option[ProcessState]] =
-    deploymentManager.findJobStatus(processIdWithName.name).recover {
+                                (implicit ec: ExecutionContext, freshnessPolicy: DataFreshnessPolicy): Future[WithDataFreshnessStatus[Option[ProcessState]]] =
+    deploymentManager.getProcessState(processIdWithName.name).recover {
```
Comment (gadomsky) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala:270]: we do not need `WithDataFreshnessStatus` here really?
can we do `deploymentManager.getProcessState(processIdWithName.name).map(_.value)` ?

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DeploymentManager.scala (hunk 2 above)
Comment (arkadius) [reply designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DeploymentManager.scala:48]: good idea

**Review by gadomsky — APPROVED** [review]
LGTM, thanks
